crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -153,30 +154,49 @@ impl BlobVec {
     /// [`BlobVec`]'s `item_layout`
     pub unsafe fn replace_unchecked(&mut self, index: usize, value: OwningPtr<'_>) {
         debug_assert!(index < self.len());
-        // If `drop` panics, then when the collection is dropped during stack unwinding, the
-        // collection's `Drop` impl will call `drop` again for the old value (which is still stored
-        // in the collection), so we get a double drop. To prevent that, we set len to 0 until we're
-        // done.
+
+        let drop = self.drop;
+        let size = self.item_layout.size();
+
+        // Temporarily set the length to zero, so that uninitialized bytes won't
+        // get observed in case this function panics.
         let old_len = self.len;
-        let ptr = self.get_unchecked_mut(index).promote().as_ptr();
         self.len = 0;
```

Comment:
Thinking on this a bit more. This doesn't need to be done unless the drop value is Some.

Perhaps we should just branch on - drop and simplify the non-drop path.

Reply:
I was doubtful at first but this suggestion turned out very nicely.